### PR TITLE
adding support for builder decorator

### PR DESCRIPTION
**Pull request by srohatgi**
This commit adds support for a custom decorator to be specified for a Builder. This feature may be used to add client instrumentation and/ or aggregation.

### CODE REVIEW

**Comment by gaurav-chawla** [comment]
Can you please update the example_test.go?

File: graph.go [hunk 1]
```diff
@@ -40,23 +40,30 @@ import (
 // Opts captures customizable functionality like logging
 type Opts struct {
 	CustomLogger func(args ...interface{})
+	Decorator    func(b Builder) Builder
```
Comment (gaurav-chawla) [inline graph.go]: Shouldn't this be like `func(ctx context.Context, b graph.Builder) (string, error) ` and the client will call the update instead the library?

**Comment by srohatgi** [comment]
@gaurav-chawla checkout https://cd.splunkdev.com/core-services/maestro-tenant-controller/merge_requests/803 for an example

**Comment by gaurav-chawla** [comment]
Looks good.

**Review by gaurav-chawla — APPROVED** [review]

**Pull request merged**